extractor/docker/docker.go

```diff
@@ -257,29 +311,138 @@ func (d Extractor) extractLayerFiles(layerCh chan layer, errCh chan error, ctx c
 	return nil
 }
 
-func (d Extractor) extractLayerWorker(dig digest.Digest, r *registry.Registry, ctx context.Context, image registry.Image, errCh chan error, layerCh chan layer) {
-	var rc io.Reader
-	// Use cache
-	rc = d.Cache.Get(string(dig))
-	if rc == nil {
-		// Download the layer.
-		layerRC, err := r.DownloadLayer(ctx, image.Path, dig)
+func (d Extractor) extractLayerWorker(dig digest.Digest, r *registry.Registry, ctx context.Context, image registry.Image, errCh chan error, layerCh chan layer, filenames []string) {
+	var tarContent bytes.Buffer
+	var cacheContent []byte
+	found, _ := d.Cache.Get(kvtypes.GetItemInput{
+		BucketName: "layertars",
+		Key:        string(dig),
+		Value:      &cacheContent,
+	})
+
+	if found {
+		b, err := extractTarContent(cacheContent)
+		if err != nil || len(b) <= 0 {
+			found = false
+		}
+
+		_, err = tarContent.Write(b)
+		if err != nil {
+			found = false
+		}
+	}
+
+	if !found {
+		rc, err := r.DownloadLayer(ctx, image.Path, dig)
 		if err != nil {
 			errCh <- xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
 			return
 		}
+		defer rc.Close()
 
-		rc, err = d.Cache.Set(string(dig), layerRC)
+		// read the incoming gzip from the layer
+		gzipReader, err := gzip.NewReader(rc)
 		if err != nil {
-			log.Print(err)
+			errCh <- xerrors.Errorf("could not init gzip reader: %w", err)
+			return
+		}
+		defer gzipReader.Close()
+
+		tarReader := tar.NewReader(io.TeeReader(gzipReader, &tarContent))
+
+		var storeInCache bool
+		var cacheBuf bytes.Buffer
+		if len(filenames) > 0 {
+			if cacheBuf, storeInCache, err = getFilteredTarballBuffer(tarReader, filenames); err != nil {
```

Comment:
```suggestion
			if tarContent, storeInCache, err = getFilteredTarballBuffer(tarReader, filenames); err != nil {
```

Reply:
I think storeInCache is unnecessary.
```suggestion
			if cacheBuf, err = getFilteredTarballBuffer(tarReader, filenames); err != nil {
```